pkg/cmd/psa/psa.go

```diff
@@ -0,0 +1,302 @@
+				return fmt.Errorf("validation failed: %w", err)
+			}
+			if err := o.Complete(); err != nil {
+				return fmt.Errorf("completion failed: %w", err)
+			}
+
+			return o.Run()
+		},
+	}
+
+	fs := cmd.Flags()
+	o.configFlags.AddFlags(fs)
+	fs.StringVar(&o.level, "level", "restricted", "The PodSecurity level to check against.")
+	fs.BoolVar(&o.quite, "quite", false, "Do not return non-zero exit code on violations.")
+
+	return &cmd
+}
+
+// Validate ensures that all required arguments and flag values are set properly.
+func (o *PSAOptions) Validate() error {
+	if _, ok := validLevels[o.level]; !ok {
+		return fmt.Errorf("invalid level %q", o.level)
+	}
+
+	return nil
+}
+
+// Complete sets all information required for processing the command.
+func (o *PSAOptions) Complete() error {
+	config := o.configFlags.ToRawKubeConfigLoader()
+	restConfig, err := config.ClientConfig()
+	if err != nil {
+		return fmt.Errorf("failed to create rest config: %w", err)
+	}
+
+	ns, set, err := config.Namespace()
+	if err != nil {
+		return fmt.Errorf("failed to get namespace: %w", err)
+	}
+	if set {
+		o.namespace = ns
+	}
+
+	// Setup a client with a custom WarningHandler that collects the warnings.
+	o.warnings = &warningsHandler{}
+	restConfig.WarningHandler = o.warnings
+	o.client, err = kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create client: %w", err)
+	}
+
+	return nil
+}
+
+// Run attempts to update the namespace psa enforce label to the psa audit value.
+func (o *PSAOptions) Run() error {
+	// Get a list of all the namespaces.
+	namespaceList, err := o.getNamespaces()
+	if err != nil {
+		return fmt.Errorf("failed to get namespaces: %w", err)
+	}
+
+	podSecurityViolations := []*PodSecurityViolation{}
+	// Gather all the warnings for each namespace, when enforcing audit-level.
+	for _, namespace := range namespaceList.Items {
+		psv, err := o.checkNamespacePodSecurity(&namespace)
+		if err != nil {
+			return fmt.Errorf("failed to check namespace %q: %w", namespace.Name, err)
+		}
+		if psv == nil {
+			continue
+		}
+
+		klog.V(4).Infof(
+			"Pod %q has pod security violations, gathering Pod and Deployment Resources",
+			psv.PodName,
+			namespace.Name,
+		)
+
+		psv.Pod, err = o.client.CoreV1().
+			Pods(namespace.Name).
+			Get(context.Background(), psv.PodName, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf(
+				"failed to get pod %q from %q, which violates %q: %w",
+				psv.PodName,
+				namespace.Name,
+				psv.Violations[0],
+				err,
+			)
+		}
+
+		psv.PodController, err = o.getPodController(psv.Pod)
+		if err != nil {
+			return fmt.Errorf("failed to get pod controller: %w", err)
+		}
+
+		podSecurityViolations = append(podSecurityViolations, psv)
+	}
+
+	if len(podSecurityViolations) == 0 {
+		return nil
+	}
+
+	// Print the violations.
+	if err := json.NewEncoder(os.Stdout).Encode(podSecurityViolations); err != nil {
+		return err
+	}
+
+	if !o.quite {
+		os.Exit(1)
```

Comment:
huh, why isn't this only done when --quiet?

Reply:
if `quiet == false` then `!quiet == true`. So if we should not be `quiet`, we return an err exit code. Isn't this expected behavior?

Reply:
In the [lines above](https://github.com/openshift/cluster-debug-tools/pull/40/files/07ce2896145325f4b1b66ceaf3c259cd78e47d93#diff-a3d5fb2755a1063e47afb3ab8b2b1a436f5433ba352f5eebbcc41505c1ac9dc8R173-R175), we have:

```go
if len(podSecurityViolations) == 0 {
	return nil
}
```

This check ensures that we only reach the os.Exit-path if there are indeed violations detected. If two engineers independently raised concerns about this, I realize I could have done a better job.

I will review the code and consider restructuring it to make the error path and the behavior of the --quiet flag more obvious. Thank you for bringing this to my attention.

Reply:
I mean you don't have to use `os.Exit(1)`. If you have to return non zero exit code, you can just return an error and cobra translates it to exit code 1.

But in the current version of PR, if quite is true, you are returning zero exit code and for the opposite case you are returning non zero exit code which seems to be incompatible.